experimental/shared/src/main/scala/zio/prelude/experimental/BothCompose.scala

```diff
@@ -57,6 +59,32 @@ object BothCompose {
     }
 
   }
+
+  implicit val URIOApplicationCompose: ApplicationCompose[URIO] = new ApplicationCompose[URIO] {
+
+    type -->:[-t, +r] = URIO[t, r]
+
+    def application[A, B]: URIO[(URIO[A, B], A), B] = URIO.accessM[(URIO[A, B], A)] { case (a2b, a) => a2b.provide(a) }
+
+    def curry[A, B, C](f: URIO[(A, B), C]): URIO[A, URIO[B, C]] =
```

Comment:
I'm not sure that `curry` and `uncurry` is really an identity here in the presence of effects. I'm not sure if that is a guarantee you are expecting to hold.

Reply:
I know that when side effects are involved, anything goes, but I was thinking that these still might be useful instances anyway.

Reply:
Sure there is a general "anything goes" caveat but we still have a lot of properties we state hold about properties like interruption.

My concern is more specific. If you start with an effect `nested: URIO[A, URIO[B, C]]` the outer effect may perform some work, say incrementing a `ref1` and the inner effect may perform some other work, say increment `ref2`. In `f` those are separate effects and we can observe that by running the outer effect, checking the `Ref` values, and then running the inner effect. If we call `uncurry` on that we collapse it into one effect. Calling `curry` let's us pull the environment out, but we get a new nested effect where if we run the outer effect etither `Ref` will be updated and the inner effect will update both of them. So these operations aren't value preserving in a way that is pretty directly observable.

More conceptually, if an `AssociativeCompose` is a function with "something extra", `curry` and `uncurry` says we can split and recombine these function like things in a value preserving way but we actually can't do that with the "something extra" that makes `URIO[A, B]` different from `A => B`.

Reply:
Yeah, I think I see what you're getting at. `uncurry` smashes the effects together and `curry` can't "unsmash" them, it can only pull out the environment again.

Perhaps that's a good reason for not making this instance.

I'll make it just `BothCompose[URIO]`. If the need to make it anything more arises, we can add it.